### ISSUE TITLE
 server selection error: server selection timeout Error 

### ISSUE THREAD

**Issue by pcinnusamy** [rwynn/monstache#482]
Hi Rwynn

Your helping hands would be very welcome to remove the barrier at this moment :)

We have done the DocumentDB3.6 setup on our AWS cluster and created end points with replica set rs0 on primary .
Also , we have created the separate cluster for Monstache service with EC2 instance and performed connectivity testing between DocumentDB cluster and ES cluster, but we are unable to connect while running thru Monstahce configuration. PFA results for your reference.

Tring to connect DocumentDB from Monstahce cluster via telnet :(connection between db is open)
]# telnet mongodb-poc-cluster-***.amazonaws.com 27017
Trying ipaddress.
Connected to mongodb-poc-cluster-.docdb.amazonaws.com.
Escape character is '^]'.

Steps followed for Monstache installation :
1.wget https://dl.google.com/go/go1.14.4.linux-amd64.tar.gz
2. tar -C /usr/local -xzf go1.14.4.linux-amd64.tar.gz
3. Configure environment variables.
Run the vim /etc/profile command to open the configuration file for environment variables. Then, add the following content to the file. GOPROXY specifies a proxy for the modules of Alibaba Cloud SDK for Go.
export GOROOT=/usr/local/go
export GOPATH=/home/go/
export PATH=$PATH:$GOROOT/bin:$GOPATH/bin
4.Apply the environment variables.
source /etc/profile
Go to the installation path.
cd /usr/local/
5. Download the installation package from the GitHub repository.
git clone https://github.com/rwynn/monstache.git
6. Go to the monstache directory.
cd monstache
7. git checkout rel6
8.Install Monstache.
go install
9. monstache -v
6.7.4
10. cd /usr/local/monstache/
vim config.toml
11. Modify the configuration file.
PFA file.
12.After placed configuration file , Run Monstache.
monstache -f config.toml

Executing Monstache service for accessing DocumentDB :
[root@ip- /doc/poc/monstache/monstache]# monstache -f config.toml
ERROR 2021/02/18 12:07:49 Unable to connect to MongoDB using URL mongodb://REDACTED@mongodb-poc-cluster.cluster-*******.amazonaws.com:27017/?authSource=admin&replicaSet=rs0&ssl_ca_certs=rds-combined-ca-bundle.pem: server selection error: server selection timeout, current topology: { Type: ReplicaSetNoPrimary, Servers: [{ Addr: mongodb-poc-cluster.cluster-chxlzpkqm5yc.us-east-1.docdb.amazonaws.com:27017, Type: Unknown, State: Connected, Average RTT: 0, Last error: connection() : connection(mongodb-poc-cluster.cluster-**********.amazonaws.com:27017[-5]) incomplete read of message header: read tcp 10.31.35.125:39722->10.31.9.164:27017: i/o timeout }, ] }

Executing Monstache service for accessing MongoDB :( working without any issue )
Configuration :
config.toml

mongo-url="mongodb://mongopoc_admin:@mongodb-poc-cluster.cluster-*************.amazonaws.com:27017/?authSource=admin&replicaSet=rs0&ssl_ca_certs=rds-combined-ca-bundle.pem"
elasticsearch-urls = ["https://:9200"]
direct-read-namespaces = ["admin.Docdbtoes"]
change-stream-namespaces = ["admin.Docdbtoes"]
gzip = true
stats = true
index-stats = true
elasticsearch-user = "poc-documentdb"
elasticsearch-password = "****"
elasticsearch-max-conns = 4
elasticsearch-pem-file = "/doc/poc/monstache/monstache/pem/es.pem"
elastic-validate-pem-file = true
dropped-collections = true
dropped-databases = true
resume = true
resume-strategy = 0
index-files = true
exit-after-direct-reads =true

**Comment by rwynn** [comment]
Hi @pcinnusamy monstache passes the connection URI directly through to the MongoDB go driver. You can find details about how to configure that URI for the driver at [Configuring the MongoDB Go Driver connection URI](https://pkg.go.dev/go.mongodb.org/mongo-driver/mongo#Connect)

Also, I have some configurations documented for your specific use case at [Configuring Monstache for DocumentDB](https://rwynn.github.io/monstache-site/advanced/#amazon-documentdb-with-mongodb-compatibility).  

If you are able to get a simple connection and Ping call to succeed as documented [here](https://github.com/mongodb/mongo-go-driver#usage) then you should be able to use the same URI with Monstache and get the same result.

**Comment by rwynn** [comment]
@pcinnusamy for URI params you may need to use `ssl=true` and `sslcertificateauthorityfile=rds-combined-ca-bundle.pem` per https://docs.aws.amazon.com/documentdb/latest/developerguide/connect_programmatically.html#connect_programmatically-tls_enabled.  

See https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/connstring/connstring.go for the full list.

I believe you will also need `readPreference=primary` for change streams according to Limitations at https://docs.aws.amazon.com/documentdb/latest/developerguide/change_streams.html.  

This is documented at https://rwynn.github.io/monstache-site/advanced/#amazon-documentdb-with-mongodb-compatibility but it's easy to miss.

So in summary maybe try...
```
...?authSource=admin&replicaSet=rs0&ssl=true&sslcertificateauthorityfile=rds-combined-ca-bundle.pem&readPreference=primary
```

**Comment by pcinnusamy** [comment]
Thank you so much for your valuable input Rywn !!!!
Awesome :):):) it got successfully connected and it's working fine now !!!!

[root@ip-0.0.0.0 /doc/poc/monstache/monstache]# monstache -f config.toml
INFO 2021/02/20 03:27:21 Started monstache version 6.7.4
INFO 2021/02/20 03:27:21 Go version go1.14.4
INFO 2021/02/20 03:27:21 MongoDB go driver v1.4.6
INFO 2021/02/20 03:27:21 Elasticsearch go driver 7.0.22
**INFO 2021/02/20 03:27:21 Successfully connected to MongoDB version 4.0.0**

**Comment by rwynn** [comment]
@pcinnusamy glad to hear that!  Looks like DocumentDB is advertising as MongoDB 4 now.  That is interesting.  

When you get a chance can you comment here with the final settings that worked so others might benefit from this issue?

**Comment by pcinnusamy** [comment]
Hi Rwynn,

Yes , AWS has recently announced that Document DB  having feature of MongoDB 4.0  version, that's why we are also migrating from 3.6 to 4.0.  Also I will post end to end setup details and hurdles which I faced once completed this migration.

After this execution I am getting an error while connecting elastic server with given configuration, please suggest what needs to be done on this. 

Note : I just took  issue as reference and done  my set up , even though its not working. 

Unable to create Elasticsearch client: health check timeout: no Elasticsearch node available #250

Configuration :
**mongo-url="mongodb://mongopoc_admin:password@mongodb-************.amazonaws.com:27017/??authSource=admin&replicaSet=rs0&ssl=true&sslcertificateauthorityfile=rds-combined-ca-bundle.pem&readPreference=primary"
elasticsearch-urls = ["https://es-abced.io:9200"]
direct-read-namespaces = ["admin.Docdbtoes"]
change-stream-namespaces = ["admin.Docdbtoes"]
gzip = true
stats = true
index-stats = true
elasticsearch-user = "poc-docudb"
elasticsearch-password = "*******************"**
elasticsearch-max-conns = 4
elasticsearch-pem-file = "/doc/poc/monstache/pem/es.pem"
elasticsearch-validate-pem-file = false
elasticsearch-healthcheck-timeout-startup = 200
elasticsearch-healthcheck-timeout = 200
dropped-collections = true
dropped-databases = true
resume = true
resume-strategy = 0
index-files = true
exit-after-direct-reads =true

**Via telnet command able to access ES cluster :**
[root@myip /doc/poc/monstache/monstache]# telnet es-abced.io 9200
Trying 10.10.1.159...
Connected to es-abced.io.
Escape character is '^]'.


**Via Monstahce script :**
Error : [root@myipaddress /doc/poc/monstache/monstache]# monstache -f config.toml
INFO 2021/02/20 03:38:27 Started monstache version 6.7.4
INFO 2021/02/20 03:38:27 Go version go1.14.4
INFO 2021/02/20 03:38:27 MongoDB go driver v1.4.6
INFO 2021/02/20 03:38:27 Elasticsearch go driver 7.0.22
INFO 2021/02/20 03:38:27 Successfully connected to MongoDB version 4.0.0
**ERROR 2021/02/20 03:41:48 Unable to create Elasticsearch client: health check timeout: no Elasticsearch node available**

**Comment by rwynn** [comment]
What do you get when you `curl "https://es-abced.io:9200"` from this host ?

**Comment by pcinnusamy** [comment]
Hi Rywnn,

The below response I am getting when executing curl "https://es-abced.io:9200" command 

[root@myip /doc/poc/monstache]# curl  "https://es-abced.io:9200"
curl: (52) Empty reply from server

**Comment by rwynn** [comment]
Actually, you might need to try something like this...
```
curl --cacert /doc/poc/monstache/pem/es.pem "https://es-abced.io:9200" --user poc-docudb:password
```

**Comment by pcinnusamy** [comment]
It didn't work Rwynn :(

[root@myip /doc/poc/monstache]/# curl --cacert /doc/poc/monstache/pem/es.pem "https://es-abced.io:9200" --user poc-docudb:password
curl: (35) error:0407008A:rsa routines:RSA_padding_check_PKCS1_type_1:invalid padding

[root@myip /doc/poc/monstache]/# curl --cacert /doc/poc/monstache/pem/es.pem "es-abced.io:9200" --user poc-docudb:password
curl: (52) Empty reply from server
[root@myip /doc/poc/monstache]#
